File: 904. Fruit Into Baskets/solution.py
class Solution(object):
    def totalFruit(self, tree):
        """
        :type tree: List[int]
        :rtype: int
        """
        size = 0
        if not tree:
            return size
        
        seen = set()
        count = 0
        prev = None
        pcount = 0
        for fruit in tree:
            if fruit in seen:
                if fruit == prev:
                    pcount += 1
                else:
                    pcount = 1
                    prev = fruit
            else:
                if len(seen) == 2:
                    count = pcount
                    for f in seen:
                        if f != prev:
                            seen.remove(f)
                            break
                seen.add(fruit)
                prev = fruit
                pcount = 1
            count += 1
            size = max(size, count)
        return max(size, count)
                
                

File: 904. Fruit Into Baskets/solution.java
class Solution {
    public int totalFruit(int[] tree) {
        // edge cases
        int n = tree.length;
        if (n < 3) {
            return n;
        }

        // maintain a basket set for each window
        HashSet<Integer> basket = new HashSet<Integer>();
        int prevFruit = -1;
        int prevFruitCount = 0;
        int count = 0;
        int maxCount = 0;
        int fruit;
        for (int i = 0; i < n; i++) {
            fruit = tree[i];
            if (basket.contains(fruit)) {
                count++;
                if (prevFruit != fruit) {
                    prevFruitCount = 0;
                    prevFruit = fruit;
                } 
                prevFruitCount++;
            } else {
                maxCount = Math.max(maxCount, count);
                if (basket.size() == 2) {
                    basket.clear();
                    basket.add(prevFruit);
                } 
                basket.add(fruit);
                count = prevFruitCount + 1;
                prevFruit = fruit;
                prevFruitCount = 1;
            }
        }
        maxCount = Math.max(maxCount, count);
        return maxCount;
    }
}
